### DISCUSSION TITLE
[Q&A] AttributeError: module 'cryptography.hazmat.primitives.serialization' has no attribute 'pkcs12'

### DISCUSSION THREAD

**Discussion by hwpang** [NVIDIA/NVFlare#2741] [Q&A] [answered]
### Python version (`python3 -V`)

3.10.14

### NVFlare version (`python3 -m pip list | grep "nvflare"`)

2.4.1

### NVFlare branch (if running examples, please use the branch that corresponds to the NVFlare version, `git branch`)

2.4

### Operating system

Ubuntu 20.04

### Have you successfully run any of the following examples?

- [X] hello-numpy-sag with simulator
- [X] hello-pt with simulator
- [ ] hello-numpy-sag with POC
- [ ] hello-pt with POC

### Please describe your question

I'm a new user of NVFlare and am currently going through the tutorials. I am able to run FL in the simulation mode, but I'm not able to get POC mode to work.

When I run `nvflare poc prepare -n 2`, I had the following error:

```
prepare poc at /tmp/nvflare/poc for 2 clients
This will delete poc workspace directory: '/tmp/nvflare/poc' and create a new one. Is it OK to proceed? (y/N) y
provision at /tmp/nvflare/poc for 2 clients with /tmp/nvflare/poc/project.yml
Exception raised during provision.  Incomplete prod_n folder removed.
Traceback (most recent call last):
  File "/anaconda/envs/nvflare-env/lib/python3.10/site-packages/nvflare/lighter/spec.py", line 174, in provision
    b.build(project, ctx)
  File "/anaconda/envs/nvflare-env/lib/python3.10/site-packages/nvflare/lighter/impl/cert.py", line 137, in build
    self._build_write_cert_pair(server, "server", ctx)
  File "/anaconda/envs/nvflare-env/lib/python3.10/site-packages/nvflare/lighter/impl/cert.py", line 119, in _build_write_cert_pair
    pkcs12 = serialization.pkcs12.serialize_key_and_certificates(
AttributeError: module 'cryptography.hazmat.primitives.serialization' has no attribute 'pkcs12'
```

However, if I open up a Python REPL, I can import `pkcs12` from `cryptography.hazmat.primitives.serialization`:

```
Python 3.10.14 (main, May  6 2024, 19:42:50) [GCC 11.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from cryptography.hazmat.primitives.serialization import pkcs12
>>> 
```

**Comment by YuanTingHsieh** [comment]
@hwpang thanks for reporting.

@IsaacYangSLA do we need to also apply this https://github.com/NVIDIA/NVFlare/pull/2721 to 2.4?

**Reply by YuanTingHsieh** [reply, marked as answer]
Should be fixed by https://github.com/NVIDIA/NVFlare/pull/2746

**Reply by hwpang** [reply]
Thanks! It is fixed! I got the expected output after https://github.com/NVIDIA/NVFlare/pull/2746.

```
nvflare poc prepare -n 2
prepare poc at /tmp/nvflare/poc for 2 clients
This will delete poc workspace directory: '/tmp/nvflare/poc' and create a new one. Is it OK to proceed? (y/N) y
provision at /tmp/nvflare/poc for 2 clients with /tmp/nvflare/poc/project.yml
Generated results can be found under /tmp/nvflare/poc/example_project/prod_00. 
```
